File: CS/Q212620/Northwind.cs
using System;
using DevExpress.Xpo;
using DevExpress.XtraTreeList;
using System.Collections.Generic;
using System.Collections;
namespace Northwind {

    public class Categories : XPVirtualTreeListDataBase {
        int fCategoryID;
        [Key(true)]
        public int CategoryID {
            get { return fCategoryID; }
            set { SetPropertyValue<int>("CategoryID", ref fCategoryID, value); }
        }
        string fCategoryName;
        [Size(15)]
        public string CategoryName {
            get { return fCategoryName; }
            set { SetPropertyValue<string>("CategoryName", ref fCategoryName, value); }
        }
        [Association("Category-Products", typeof(Products))]
        public XPCollection<Products> Products { get { return GetCollection<Products>("Products"); } }
        protected override string ChildrenCollectionName { get { return "Products"; } }
        public Categories(Session session) : base(session) { }
        public Categories() : this(Session.DefaultSession) { }
    }

    public class Products : XPVirtualTreeListDataBase {
        int fProductID;
        [Key(true)]
        public int ProductID {
            get { return fProductID; }
            set { SetPropertyValue<int>("ProductID", ref fProductID, value); }
        }
        string fProductName;
        [Size(40)]
        public string ProductName {
            get { return fProductName; }
            set { SetPropertyValue<string>("ProductName", ref fProductName, value); }
        }
        Categories fCategoryID;
        [Association("Category-Products")]
        public Categories CategoryID {
            get { return fCategoryID; }
            set { SetPropertyValue<Categories>("CategoryID", ref fCategoryID, value); }
        }
        public Products(Session session) : base(session) {
            fieldMap.Add("CategoryName", "ProductName");
        }
        public Products() : this(Session.DefaultSession) { }
    }

    [NonPersistent]
    public class XPVirtualTreeListDataBase : XPLiteObject, TreeList.IVirtualTreeListData {
        protected XPCollection childrenCore;
        protected Dictionary<string, string> fieldMap = new Dictionary<string, string>();

        public XPVirtualTreeListDataBase() : base() { }

        public XPVirtualTreeListDataBase(Session session) : base(session) { }

        protected virtual string ChildrenCollectionName { get { return string.Empty; } }

        private string GetTreeListDataFieldName(string fieldName) {
            return fieldMap.ContainsKey(fieldName) ? fieldMap[fieldName] : fieldName;
        }

        public void SetChildren(XPCollection children) {
            childrenCore = children;
        }

        #region IVirtualTreeListData Members

        public void VirtualTreeGetCellValue(VirtualTreeGetCellValueInfo info) {
            string fieldName = GetTreeListDataFieldName(info.Column.FieldName);
            info.CellData = ClassInfo.GetMember(fieldName).GetValue(this);
        }

        public void VirtualTreeGetChildNodes(VirtualTreeGetChildNodesInfo info) {
            if (string.IsNullOrEmpty(ChildrenCollectionName)) info.Children = childrenCore;
            else info.Children = (IList)ClassInfo.GetMember(ChildrenCollectionName).GetValue(this);
        }

        public void VirtualTreeSetCellValue(VirtualTreeSetCellValueInfo info) {
            string fieldName = GetTreeListDataFieldName(info.Column.FieldName);
            ClassInfo.GetMember(fieldName).SetValue(this, info.NewCellData);
        }

        #endregion
    }
}

File: VB/Q212620/Northwind.vb
Imports DevExpress.Xpo
Imports DevExpress.XtraTreeList
Imports System.Collections.Generic
Imports System.Collections

Namespace Northwind

    Public Class Categories
        Inherits XPVirtualTreeListDataBase

        Private fCategoryID As Integer

        <Key(True)>
        Public Property CategoryID As Integer
            Get
                Return fCategoryID
            End Get

            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("CategoryID", fCategoryID, value)
            End Set
        End Property

        Private fCategoryName As String

        <Size(15)>
        Public Property CategoryName As String
            Get
                Return fCategoryName
            End Get

            Set(ByVal value As String)
                SetPropertyValue("CategoryName", fCategoryName, value)
            End Set
        End Property

        <Association("Category-Products", GetType(Products))>
        Public ReadOnly Property Products As XPCollection(Of Products)
            Get
                Return GetCollection(Of Products)("Products")
            End Get
        End Property

        Protected Overrides ReadOnly Property ChildrenCollectionName As String
            Get
                Return "Products"
            End Get
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Sub New()
            Me.New(Session.DefaultSession)
        End Sub
    End Class

    Public Class Products
        Inherits XPVirtualTreeListDataBase

        Private fProductID As Integer

        <Key(True)>
        Public Property ProductID As Integer
            Get
                Return fProductID
            End Get

            Set(ByVal value As Integer)
                SetPropertyValue(Of Integer)("ProductID", fProductID, value)
            End Set
        End Property

        Private fProductName As String

        <Size(40)>
        Public Property ProductName As String
            Get
                Return fProductName
            End Get

            Set(ByVal value As String)
                SetPropertyValue("ProductName", fProductName, value)
            End Set
        End Property

        Private fCategoryID As Categories

        <Association("Category-Products")>
        Public Property CategoryID As Categories
            Get
                Return fCategoryID
            End Get

            Set(ByVal value As Categories)
                SetPropertyValue("CategoryID", fCategoryID, value)
            End Set
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
            fieldMap.Add("CategoryName", "ProductName")
        End Sub

        Public Sub New()
            Me.New(Session.DefaultSession)
        End Sub
    End Class

    <NonPersistent>
    Public Class XPVirtualTreeListDataBase
        Inherits XPLiteObject
        Implements TreeList.IVirtualTreeListData

        Protected childrenCore As XPCollection

        Protected fieldMap As Dictionary(Of String, String) = New Dictionary(Of String, String)()

        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Protected Overridable ReadOnly Property ChildrenCollectionName As String
            Get
                Return String.Empty
            End Get
        End Property

        Private Function GetTreeListDataFieldName(ByVal fieldName As String) As String
            Return If(fieldMap.ContainsKey(fieldName), fieldMap(fieldName), fieldName)
        End Function

        Public Sub SetChildren(ByVal children As XPCollection)
            childrenCore = children
        End Sub

#Region "IVirtualTreeListData Members"
        Public Sub VirtualTreeGetCellValue(ByVal info As VirtualTreeGetCellValueInfo) Implements TreeList.IVirtualTreeListData.VirtualTreeGetCellValue
            Dim fieldName As String = GetTreeListDataFieldName(info.Column.FieldName)
            info.CellData = ClassInfo.GetMember(fieldName).GetValue(Me)
        End Sub

        Public Sub VirtualTreeGetChildNodes(ByVal info As VirtualTreeGetChildNodesInfo) Implements TreeList.IVirtualTreeListData.VirtualTreeGetChildNodes
            If String.IsNullOrEmpty(ChildrenCollectionName) Then
                info.Children = childrenCore
            Else
                info.Children = CType(ClassInfo.GetMember(ChildrenCollectionName).GetValue(Me), IList)
            End If
        End Sub

        Public Sub VirtualTreeSetCellValue(ByVal info As VirtualTreeSetCellValueInfo) Implements TreeList.IVirtualTreeListData.VirtualTreeSetCellValue
            Dim fieldName As String = GetTreeListDataFieldName(info.Column.FieldName)
            ClassInfo.GetMember(fieldName).SetValue(Me, info.NewCellData)
        End Sub
#End Region
    End Class
End Namespace
